evm/src/cpu/kernel/asm/core/terminate.asm

```diff
@@ -0,0 +1,56 @@
+// Handlers for operations which terminate the current context, namely STOP,
+// RETURN, SELFDESTRUCT, REVERT, and exceptions such as stack underflow.
+
+global stop:
+    // TODO: Set parent context's CTX_METADATA_RETURNDATA_SIZE to 0.
+    %jump(terminate_common)
+
+global return:
+    // TODO: Set parent context's CTX_METADATA_RETURNDATA_SIZE.
+    // TODO: Copy returned memory to parent context's RETURNDATA
+    // TODO: Copy returned memory to parent context's memory (as specified in their call instruction)
+    %jump(terminate_common)
+
+global selfdestruct:
+    %consume_gas_const(@GAS_SELFDESTRUCT)
+    // TODO
+    %jump(terminate_common)
+
+global revert:
+    // TODO
+    %jump(terminate_common)
+
+// The execution is in an exceptional halt-ing state if
+// - there is insufficient gas
+// - the instruction is invalid
+// - there are insufficient stack items
+// - a JUMP/JUMPI destination is invalid
+// - the new stack size would be larger than 1024, or
+// - state modification is attempted during a static call
+global exception:
+    // TODO
+    %jump(terminate_common)
+
+terminate_common:
```

Comment:
Do we want to charge gas here or has that been done?

Reply:
Like charge the cost of these terminal instructions? It looks like `STOP`, `RETURN` and `REVERT` are free (in fact the only free ones). The specs do have a "zero gas" constant for them, so we could do `%consume_gas_const(@GAS_ZERO)` and optimize it out if we wanted to match.

Or charge gas for the transaction? I believe that's done at the start, and then leftover gas is refunded upon graceful termination (just not exceptions). I'll add a TODO to refund gas to the parent context upon `STOP`, `RETURN` and `REVERT` or `SELFDESTRUCT`.